### PR TITLE
[EBPF] Add BeforeStop/AfterStop method to modifiers

### PR DESCRIPTION
**Pull request by gjulianm**
### What does this PR do?

This PR adds a `BeforeStop` method to the `Modifier` interface, so that modifiers can clean up their structures before the manager stops.

This is a re-make of https://github.com/DataDog/datadog-agent/pull/32414 with an alternative implementation, similar to what @brycekahle was implementing in https://github.com/DataDog/datadog-agent/pull/31402/files

### Motivation

The telemetry modifier needs to clean up some structures so that the same program can be unloaded and reloaded multiple times. The `shared-libraries` program was having problems with this case.

### Describe how you validated your changes

Linting/building/unit tests passing.

### CODE REVIEW

File: pkg/ebpf/manager.go [hunk 1]
```diff
@@ -81,22 +131,47 @@ func (m *Manager) InitWithOptions(bytecode io.ReaderAt, opts *manager.Options) e
...
+// Stop is a wrapper around ebpf-manager.Manager.Stop
+func (m *Manager) Stop(cleanupType manager.MapCleanupType) error {
+	var errs error
+
 	for _, mod := range m.EnabledModifiers {
-		log.Tracef("Running %s manager modifier AfterInit", mod)
-		if err := mod.AfterInit(m.Manager, m.Name, opts); err != nil {
-			return fmt.Errorf("error running %s manager modifier: %w", mod, err)
+		log.Tracef("Running %s manager modifier BeforeStop", mod)
+		if as, ok := mod.(ModifierBeforeStop); ok {
```
Comment (usamasaqib) [inline pkg/ebpf/manager.go:156]: Why not use `runModifierOfType`?

File: pkg/ebpf/manager.go [hunk 2]
```diff
@@ -81,22 +131,47 @@ func (m *Manager) InitWithOptions(bytecode io.ReaderAt, opts *manager.Options) e
...
 	}
-	return nil
+
+	if err := m.Manager.Stop(cleanupType); err != nil {
+		errs = errors.Join(errs, fmt.Errorf("failed to stop manager %w", err))
+	}
+
+	for _, mod := range m.EnabledModifiers {
+		log.Tracef("Running %s manager modifier AfterStop", mod)
+		if as, ok := mod.(ModifierAfterStop); ok {
```
Comment (usamasaqib) [inline pkg/ebpf/manager.go:169]: Why not use `runModifierOfType`?

File: pkg/ebpf/manager.go (hunk 1 above)
Comment (gjulianm) [reply pkg/ebpf/manager.go:156]: Overlook, changed!

**Review by usamasaqib — APPROVED** [review]

**Pull request merged**